File: C1.UWP.FlexGrid/CS/FlexGridSamples/CellFactory/TemplateCell.cs
using C1.Xaml.FlexGrid;
using System;
using System.Reflection;
using Windows.UI.Text;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;
using Windows.UI.Xaml.Controls.Primitives;
using Windows.UI.Xaml.Data;

namespace FlexGridSamples
{
    public abstract class TemplateCell : StackPanel
    {
        private static ResourceDictionary _customTemplatesDictionary;
        public static ResourceDictionary CustomTemplatesDictionary
        {
            get
            {
                if (_customTemplatesDictionary == null)
                {
                    _customTemplatesDictionary = new ResourceDictionary();
                    if (typeof(TemplateCell).GetTypeInfo().Module.Name.EndsWith("exe"))
                    {
                        _customTemplatesDictionary.Source = new Uri("ms-appx:///Resources/CustomTemplatesDictionary.xaml");
                    }
                    else
                    {
                        _customTemplatesDictionary.Source = new Uri("ms-appx:///FlexGridSamplesLib/Resources/CustomTemplatesDictionary.xaml");
                    }
                }
                return _customTemplatesDictionary;
            }
        }

        protected ContentControl Icon { get; set; }
        protected TextBlock TextBlock { get; set; }

        private Row Row { get; set; }

        public TemplateCell(Row row)
        {
            Row = row;
            Orientation = Orientation.Horizontal;

            Icon = new ContentControl();
            Icon.Width = 25;
            Icon.Height = 25;
            Icon.VerticalAlignment = VerticalAlignment.Center;
            Icon.Margin = new Thickness(4);
            Icon.ContentTemplate = GetIcon();
            Children.Add(Icon);

            TextBlock = new TextBlock();
            TextBlock.VerticalAlignment = VerticalAlignment.Center;
            Children.Add(TextBlock);
            BindCell(row.DataItem);
        }

        void BindCell(object dataItem)
        {
            var binding = new Binding { Path = new PropertyPath("Name") };
            binding.Source = dataItem;
            TextBlock.SetBinding(TextBlock.TextProperty, binding);
            TextBlock.Foreground = Row.Grid.Foreground;
        }
        public abstract DataTemplate GetIcon();

    }

    /// <summary>
    /// Represents a grid cell that is bound to a song name.
    /// </summary>
    public class SongCell : TemplateCell
    {
        public SongCell(Row row) : base(row) { }
        public override DataTemplate GetIcon()
        {
            return TemplateCell.CustomTemplatesDictionary["SongIcon"] as DataTemplate;
        }
    }

    /// <summary>
    /// Represents a grid cell that has a collapse/expand icon, an image, and some text.
    /// </summary>
    public abstract class NodeCell : TemplateCell
    {
        GroupRow _gr;
        private ToggleButton _nodeExpandButton;

        public NodeCell(Row row)
            : base(row)
        {
            // store reference to row
            _gr = row as GroupRow;
            // initialize collapsed/expanded image
            _nodeExpandButton = new ToggleButton();
            if (Util.IsWindowsPhoneDevice())
            {
                _nodeExpandButton.Template = TemplateCell.CustomTemplatesDictionary["ExpandCollapsePhoneTemplate"] as ControlTemplate;
            }
            else
            {
                _nodeExpandButton.Template = TemplateCell.CustomTemplatesDictionary["ExpandCollapseTemplate"] as ControlTemplate;
            }

            _nodeExpandButton.IsChecked = !_gr.IsCollapsed;
            _nodeExpandButton.Checked += OnExpandButtonToggled;
            _nodeExpandButton.Unchecked += OnExpandButtonToggled;
            Children.Insert(0, _nodeExpandButton);

            // make text bold
            TextBlock.FontWeight = FontWeights.Bold;
        }

        void OnExpandButtonToggled(object sender, RoutedEventArgs e)
        {
            var toggleButton = sender as ToggleButton;
            _gr.IsCollapsed = !toggleButton.IsChecked.Value;
        }
    }

    /// <summary>
    /// Represents a grid cell that is bound to an artist.
    /// </summary>
    public class ArtistCell : NodeCell
    {
        public ArtistCell(Row row)
            : base(row)
        {
        }
        public override DataTemplate GetIcon()
        {
            return TemplateCell.CustomTemplatesDictionary["ArtistIcon"] as DataTemplate;
        }
    }

    /// <summary>
    /// Represents a grid cell that is bound to an album.
    /// </summary>
    public class AlbumCell : NodeCell
    {
        public AlbumCell(Row row)
            : base(row)
        {
        }
        public override DataTemplate GetIcon()
        {
            return TemplateCell.CustomTemplatesDictionary["ArtistIcon"] as DataTemplate;
        }
    }

}

File: C1.UWP.FlexGrid/VB/FlexGridSamples/CellFactory/TemplateCell.vb
Imports Windows.UI.Xaml.Data
Imports Windows.UI.Xaml.Controls.Primitives
Imports Windows.UI.Xaml.Controls
Imports Windows.UI.Xaml
Imports Windows.UI.Text
Imports System.Reflection
Imports System
Imports C1.Xaml.FlexGrid

Public MustInherit Class TemplateCell
    Inherits StackPanel
    Private Shared _customTemplatesDictionary As ResourceDictionary
    Public Shared ReadOnly Property CustomTemplatesDictionary() As ResourceDictionary
        Get
            If _customTemplatesDictionary Is Nothing Then
                _customTemplatesDictionary = New ResourceDictionary()
                If GetType(TemplateCell).GetTypeInfo().[Module].Name.EndsWith("exe") Then
                    _customTemplatesDictionary.Source = New Uri("ms-appx:///Resources/CustomTemplatesDictionary.xaml")
                Else
                    _customTemplatesDictionary.Source = New Uri("ms-appx:///FlexGridSamplesLib/Resources/CustomTemplatesDictionary.xaml")
                End If
            End If
            Return _customTemplatesDictionary
        End Get
    End Property

    Protected Property Icon() As ContentControl
    Protected Property TextBlock() As TextBlock

    Private Property Row() As Row

    Public Sub New(r As Row)
        Row = r
        Orientation = Orientation.Horizontal

        Icon = New ContentControl()
        Icon.Width = 25
        Icon.Height = 25
        Icon.VerticalAlignment = VerticalAlignment.Center
        Icon.Margin = New Thickness(4)
        Icon.ContentTemplate = GetIcon()
        Children.Add(Icon)

        TextBlock = New TextBlock()
        TextBlock.VerticalAlignment = VerticalAlignment.Center
        Children.Add(TextBlock)
        BindCell(r.DataItem)
    End Sub

    Sub BindCell(dataItem As Object)
        Dim binding As New Binding() With {
            .Path = New PropertyPath("Name")
        }
        binding.Source = dataItem
        TextBlock.SetBinding(TextBlock.TextProperty, binding)
        TextBlock.Foreground = Row.Grid.Foreground
    End Sub
    Public MustOverride Function GetIcon() As DataTemplate

End Class

''' <summary>
''' Represents a grid cell that is bound to a song name.
''' </summary>
Public Class SongCell
    Inherits TemplateCell
    Public Sub New(row As Row)
        MyBase.New(row)
    End Sub
    Public Overrides Function GetIcon() As DataTemplate
        Return TryCast(TemplateCell.CustomTemplatesDictionary("SongIcon"), DataTemplate)
    End Function
End Class

''' <summary>
''' Represents a grid cell that has a collapse/expand icon, an image, and some text.
''' </summary>
Public MustInherit Class NodeCell
    Inherits TemplateCell
    Private _gr As GroupRow
    Private _nodeExpandButton As ToggleButton

    Public Sub New(row As Row)
        MyBase.New(row)
        ' store reference to row
        _gr = TryCast(row, GroupRow)
        ' initialize collapsed/expanded image
        _nodeExpandButton = New ToggleButton()
        If Util.IsWindowsPhoneDevice() Then
            _nodeExpandButton.Template = TryCast(TemplateCell.CustomTemplatesDictionary("ExpandCollapsePhoneTemplate"), ControlTemplate)
        Else
            _nodeExpandButton.Template = TryCast(TemplateCell.CustomTemplatesDictionary("ExpandCollapseTemplate"), ControlTemplate)
        End If

        _nodeExpandButton.IsChecked = Not _gr.IsCollapsed
        AddHandler _nodeExpandButton.Checked, AddressOf OnExpandButtonToggled
        AddHandler _nodeExpandButton.Unchecked, AddressOf OnExpandButtonToggled
        Children.Insert(0, _nodeExpandButton)

        ' make text bold
        TextBlock.FontWeight = FontWeights.Bold
    End Sub

    Sub OnExpandButtonToggled(sender As Object, e As RoutedEventArgs)
        Dim toggleButton As ToggleButton = TryCast(sender, ToggleButton)
        _gr.IsCollapsed = Not toggleButton.IsChecked.Value
    End Sub
End Class

''' <summary>
''' Represents a grid cell that is bound to an artist.
''' </summary>
Public Class ArtistCell
    Inherits NodeCell
    Public Sub New(row As Row)
        MyBase.New(row)
    End Sub
    Public Overrides Function GetIcon() As DataTemplate
        Return TryCast(TemplateCell.CustomTemplatesDictionary("ArtistIcon"), DataTemplate)
    End Function
End Class

''' <summary>
''' Represents a grid cell that is bound to an album.
''' </summary>
Public Class AlbumCell
    Inherits NodeCell
    Public Sub New(row As Row)
        MyBase.New(row)
    End Sub
    Public Overrides Function GetIcon() As DataTemplate
        Return TryCast(TemplateCell.CustomTemplatesDictionary("ArtistIcon"), DataTemplate)
    End Function
End Class